tests/test_execution_compute_plan.py

```diff
@@ -150,46 +143,145 @@ def test_compute_plan_single_session_failure(global_execution_env):
+    spec = factory.create_aggregate_algo()
+    aggregate_algo = sessions[0].add_aggregate_algo(spec)
+
+    # launch execution
+    previous_aggregatetuple_spec = None
+    previous_composite_traintuple_specs = []
+
+    cp_spec = factory.create_compute_plan(objective=objective)
+
+    for round_ in range(number_of_rounds):
+        # create composite traintuple on each node
+        composite_traintuple_specs = []
+        for index, dataset in enumerate(datasets):
+            kwargs = {}
+            if previous_aggregatetuple_spec:
+                kwargs = {
+                    'in_head_model_tuple': previous_composite_traintuple_specs[index],
+                    'in_trunk_model_tuple': previous_aggregatetuple_spec,
+                }
+            spec = cp_spec.add_composite_traintuple(
+                composite_algo=composite_algo,
+                dataset=dataset,
+                data_samples=[dataset.train_data_sample_keys[0 + round_]],
+                **kwargs,
+            )
+            composite_traintuple_specs.append(spec)
+
+        # create aggregate on its node
+        spec = cp_spec.add_aggregatetuple(
+            aggregate_algo=aggregate_algo,
+            worker=aggregate_worker,
+            in_models=composite_traintuple_specs,
+        )
+
+        # save state of round
+        previous_aggregatetuple_spec = spec
+        previous_composite_traintuple_specs = composite_traintuple_specs
+
+    # last round: create associated testtuple
+    for composite_traintuple_spec in previous_composite_traintuple_specs:
+        cp_spec.add_testtuple(
+            traintuple_spec=composite_traintuple_spec,
+        )
+
+    cp = sessions[0].add_compute_plan(cp_spec).future().wait()
+    tuples = (cp.list_traintuple(sessions) +
+              cp.list_composite_traintuples(sessions[0]) +
+              cp.list_aggregate_tuples(sessions[0]) +
+              cp.list_testtuples(sessions[0]))
+    for t in tuples:
+        assert t.status == 'done'
+
+
+def test_compute_plan_circular_dependency_failure(global_execution_env):
+    factory, network = global_execution_env
+    session = network.sessions[0].copy()
+
+    dataset = session.state.datasets[0]
+    objective = session.state.objectives[0]
+
+    spec = factory.create_algo()
+    algo = session.add_algo(spec)
+
+    cp_spec = factory.create_compute_plan(objective=objective)
+
+    traintuple_spec_1 = cp_spec.add_traintuple(
+        dataset=dataset,
+        algo=algo,
+        data_samples=dataset.train_data_sample_keys
+    )
+
+    traintuple_spec_2 = cp_spec.add_traintuple(
+        dataset=dataset,
+        algo=algo,
+        data_samples=dataset.train_data_sample_keys
+    )
+
+    traintuple_spec_1.in_models_ids.append(traintuple_spec_2.id)
+    traintuple_spec_2.in_models_ids.append(traintuple_spec_1.id)
+
+    with pytest.raises(substra.exceptions.InvalidRequest) as e:
+        session.add_compute_plan(cp_spec)
+
+    assert 'circular' in str(e)
```

Comment:
Not sure this will work, cf pytest documentation for usage of `pytest.raises`:

```python
    with pytest.raises(RuntimeError) as excinfo:

        def f():
            f()

        f()
    assert "maximum recursion" in str(excinfo.value)
```

Reply:
Indeed it doesn't work ([anymore!](https://github.com/pytest-dev/pytest/issues/5579)). I used `str(e.value)` like in the example above and it works: https://github.com/SubstraFoundation/substra-tests/pull/21/commits/908a719e3137f83861997df5b5a9a4b89936b738